src/Schema/SchemaDiff.php

```diff
@@ -162,6 +162,17 @@ protected function _toSql(AbstractPlatform $platform, $saveMode = false)
         }
 
         foreach ($this->changedTables as $tableDiff) {
+            // make sure to drop related orphan foreign keys before dropping indexes that are related to FKs
```

Comment:
I'm not sure how to interpret this documentation:

https://github.com/doctrine/dbal/blob/d6fd6957206e95a73ea7c77a0a7de8905d9fdc8a/src/Schema/SchemaDiff.php#L80-L84

Does it mean that foreign keys should not be dropped in this mode? As far as can tell from the code, that's the case: 

https://github.com/doctrine/dbal/blob/d6fd6957206e95a73ea7c77a0a7de8905d9fdc8a/src/Schema/SchemaDiff.php#L118-L122

If I replace `toSaveSql()` with `toSql()`, it doesn't fail with the original error (Cannot drop index '%s': needed in a foreign key constraint).

Should you just use `toSql()` instead? Otherwise, the patch seems to contradict the documentation.

Reply:
> I'm not sure how to interpret this documentation:
> Does it mean that foreign keys should not be dropped in this mode? As far as can tell from the code, that's the case:


_"following things don't happen: 3. Foreign Keys which reference tables that would otherwise be deleted."_

What I understand from the documentation is that it doesn't delete orphan foreign keys unless its reference is in use.

> 
> 
> If I replace `toSaveSql()` with `toSql()`, it doesn't fail with the original error (Cannot drop index '%s': needed in a foreign key constraint).
> 
> Should you just use `toSql()` instead? Otherwise, the patch seems to contradict the documentation.

`toSql()` (which means don't save) drops unmapped tables as well. In this case, of course, there is no problem since it drops all related FK and indexes. So this is not an option for those who wants to **save**.

Currently with the save mode, in this particular case, the generated queries results in an error. Thus it doesn't only contradict with the documentation and the command itself as well.

Reply:
Thanks for the details, @seferov. I need some time to digest this but I'm currently busy with other things. I'll try to get to this issue once I can.